### DISCUSSION TITLE
firefox: how to access downloaded files?

### DISCUSSION THREAD

**Discussion by therealkitman** [netblue30/firejail#5977] [Q&A] [answered]
Hi,
I would like to know how/where files downloaded via firejail+Firefox are stored. I downloaded a debian ISO to the Downloads folder but I can't see how to access/copy the ISO to my main system.

![image](https://github.com/netblue30/firejail/assets/14888092/36fb92f6-6e39-47a1-8669-ae2af3f58773)

1. How do I access/copy the ISO out of the sandbox to my main system? 

2. The ISO is gone if I restart firefox, no doubt due to the ephemeral nature of sandboxes. Can I make downloaded ISOs persistent until I manually delete them? 

Thanks.

**Comment by kzsa** [comment]
Hi,
Something is not set up correctly. By default, Firefox only accesses the Download folder. If I download something there, it stays there until I delete it. ("By default" I mean the normal installation.)
Which distro did you install firejail on, did you customize anything, how did you start Firefox?

Normally, after installation (and after installing any new program) you do the following:

```
~$ sudo firecfg 
```
The above command will set up profiles for you, including your Firefox profile.
Then you don't need any hocus-pocus to run firefox in the sandbox.
You can start it by clicking on a shortcut or even from the command line.

```
~$ firefox 
```

From now on, there is no difference between using Firefox "normally" and using it in a sandbox. Apart from the fact that Firefox only reaches your Download folder, of course.

If you type at the command line:
```
firejail --top
```
you can see which programs are running in the sandbox.

![image](https://github.com/netblue30/firejail/assets/59151228/bede54e2-c4d4-4c2d-b1fb-cfcf4d840560)


![image](https://github.com/netblue30/firejail/assets/59151228/b9115d3f-1774-4e53-a9bf-1bd14cdca77a)

**Comment by therealkitman** [comment]
> Which distro did you install firejail on, did you customize anything, how did you start Firefox?

Hi, I run openSUSE's tumbleweed (currently 20230816) and `firejail version 0.9.72`. I have a special .local file because I run a few instances of firefox under my $HOME folder with different profiles in a non-standard locations. I don't run `firecfg` because of that.

In this particular case I have `/home/chris/.config/firejail/firefox_iso.local` with the contents:

```
ignore apparmor
ignore noexec ${HOME}
whitelist ${HOME}/firefox
whitelist /data/chris/firefox_profiles/iso
```

I use my own .desktop file with the following:

```
[Desktop Entry]
Categories=Network;WebBrowser;GTK;
Encoding=UTF-8
Name=Firefox(Anime)
GenericName=Web Browser
Comment=Web Browser
TryExec=/home/chris/firefox/firefox
Exec=firejail --profile=/home/chris/.config/firejail/firefox_iso.local /home/chris/firefox/firefox --profile /data/chris/firefox_profiles/iso --no-remote  %u
Icon=/home/chris/firefox/browser/chrome/icons/default/default128.png
Terminal=false
StartupNotify=true
MimeType=text/html;text/xml;application/xhtml+xml;application/vnd.mozilla.xul+xml;text/mml;application/x-xpinstall;x-scheme-handler/http;x-scheme-handler/https;x-scheme-handler/ftp;
Type=Application

Actions=PrivateBrowsing;

[Desktop Action PrivateBrowsing]
Name=New Private Browsing Window
Exec=/home/chris/firefox/firefox --private-window  %u
```

**Comment by therealkitman** [comment]
I just did a quick test and added this line to my `firefox_iso.local` file -

```
whitelist ${HOME}/Downloads
```
Now files are saving as expected to that folder. I thought `${HOME}/Downloads` was whitelisted by default for the firefox profile??

**Comment by kzsa** [comment]
Hi,

If you use the pre-made profiles, then yes, the Download folder is available. You can easily check what you are accessing by using your browser as a "file manager".

![image](https://github.com/netblue30/firejail/assets/59151228/78316d95-0df4-416f-89e2-6144c1a92a2b)

<strike>But, as I understand it, you don't use those profiles.</strike> EDITED: Sorry! You are using pre-made profiles. .local just adds/overwrites parts of it. Please read these few lines, I think you will see things more clearly.

I faced similar issues when I started using firejail. I also use multiple browsers (multiple instances of the same browser), but I approached things from a slightly different angle.
The preset profiles have a lot of settings other than just allowing access to folders. 
These security settings are extremely complex, most of them (99.999%) I couldn't set myself to this level, because I don't know exactly under what circumstances some programs can work safely without being a hindrance.

My attitude is "Thank you very much for all the hard work you put into the profiles and I'll use all the ones available".

1. If something is blocking me, on the first try I try to modify the profile. (I'm not saying that it's easy to do this for all profiles with my current knowledge.)

I take a look at the available profile here:
```
/etc/firejail/<name of a program>.profile
```
I try to override the part I don't like in the location below:
```
~/.config/firejail/<name of a program>.local
```
If I do it this way, what I gain is that the complicated seccomp settings are retained, and only the part that was holding me back is changed.
[https://github.com/netblue30/firejail/wiki/Creating-overrides](url)

2. If I get stuck on something, nothing wants to work the way I want it to, I exclude it from firejail.

There are several ways to isolate something.

A)
Run with a separate firejail profile:
https://github.com/netblue30/firejail/wiki/Creating-Profiles
I don't really need case A), but here it is described in great detail how you can create a custom profile for yourself and it takes Firefox as an example.

B)
Running multiple instances in parallel with different Firefox profiles (not different firejail profiles, different Firefox profiles!)
![image](https://github.com/netblue30/firejail/assets/59151228/a7a97481-33b3-441a-af6b-9d235ca879c4)


"firefox" and "firefox-dev" do not look like two different programs from the firejail perspective. The operating system treats them the same way. Both are just "firefox" and that's it. What distinguishes them is where the executable file resides.

Here is, how I use e.g. "firefox-dev edition".

1. Installing Firefox dev (downloading, integrity checking, unpacking, ...):
```
sudo -i
rm -rf /opt/mozilla/firefox-dev && chown -R root:root /home/<username>/Downloads/temp/firefox && mv /home/<username>/Downloads/temp/firefox /opt/mozilla/firefox-dev
```
2. I make a launcer for it:
```
echo "[Desktop Entry]
Version=1.0
Type=Application
Encoding=UTF-8
Name=Firefox Dev
GenericName=Firefox Dev
Comment=Firefox Dev
Exec=firejail /opt/mozilla/firefox-dev/firefox -P "Firefox-dev" --no-remote
Terminal=false
StartupNotify=true
Icon=/opt/mozilla/firefox-dev/browser/chrome/icons/default/default128.png
Hidden=false
NoDisplay=false
Keywords=Internet;WWW;Browser;Web;Explorer;Dev;
X-MultipleArgs=false
Categories=GNOME;GTK;Network;WebBrowser;
MimeType=text/html;text/xml;application/xhtml+xml;application/xml;application/rss+xml;application/rdf+xml;image/gif;image/jpeg;image/png;x-scheme-handler/http;x-scheme-handler/https;x-scheme-handler/ftp;x-scheme-handler/chrome;video/webm;application/x-xpinstall;
StartupWMClass=FirefoxDev
Actions=new-window;new-private-window;

PrefersNonDefaultGPU=false

Name[hu]=Firefox-dev

[Desktop Action new-window]
Name=Open a New Window
Name[hu]=Új ablak nyitása
Exec=firejail /opt/mozilla/firefox-dev/firefox %u -new-window

[Desktop Action new-private-window]
Name=Open a New Private Window
Name[hu]=Új privát ablak nyitása
Exec=firejail /opt/mozilla/firefox-dev/firefox %u -private-window" > /home/<username>/.local/share/applications/Firefox-dev.desktop
```

![image](https://github.com/netblue30/firejail/assets/59151228/23655cb5-acde-445e-8057-7bfbb4ad536f)
```
firejail --top
```
![image](https://github.com/netblue30/firejail/assets/59151228/5146d1d4-c4b0-4f69-bc49-af29e5e0e0ca)

Br: Zs

**Comment by kzsa** [comment]
Off: So far I've only encountered two problems during system-wide use (firecfg way).
1. The gnome calculator would not start (partially solved).
2. vscodium's profile is too restrictive for me, it holds me back on some things, it's hard to find a compromise between security and usability.
I still prefer to use firecfg, because I think I gain a lot with it.

**Comment by therealkitman** [comment]
HI kzsa, thanks for the replies. I read them quickly but will re-read tomorrow after I get some sleep :)

But I still can't find why `Downloads` was "weird" using the pre-made `/etc/firejail/firefox.profile` which included `/etc/firejail/firefox-common.profile` which has the default whitelisting of `whitelist ${DOWNLOADS}` and I had to manually add that to my local file.

**Comment by ghost** [comment]
Just a quick note on `${DOWNLOADS}`. Firejail follows the `XDG` specs and in this context files like ~/.config/user-dirs.dirs and ~/.config/user-dirs.locale are important ones to actually take into account translating /home/foo/Downloads into ${DOWNLOADS} behind the scenes. A user can safely opt to change defaults for any of these XDG_XXX_DIR variables and for example use /home/foo/bar as ${DOWNLOADS}. You might want to check what these files contain on your machine. The ~/.config/user-dirs.locale file might relate to [discussion 5974](https://github.com/netblue30/firejail/discussions/5974). For more detailed info on the topic, see https://wiki.archlinux.org/title/XDG_user_directories

**Comment by therealkitman** [comment]
> Just a quick note on `${DOWNLOADS}`.

Hi @glitsj16 . I checked my `/home/chris/.config/user-dirs.dirs` file and it contains the DOWNLOADS variable so I guess that aspect is ok.

```
# This file is written by xdg-user-dirs-update
# If you want to change or add directories, just edit the line you're
# interested in. All local changes will be retained on the next run.
# Format is XDG_xxx_DIR="$HOME/yyy", where yyy is a shell-escaped
# homedir-relative path, or XDG_xxx_DIR="/yyy", where /yyy is an
# absolute path. No other format is supported.
# 
XDG_DESKTOP_DIR="$HOME/Desktop"
XDG_DOWNLOAD_DIR="$HOME/Downloads"
XDG_TEMPLATES_DIR="$HOME/Templates"
XDG_PUBLICSHARE_DIR="$HOME/Public"
XDG_DOCUMENTS_DIR="$HOME/Documents"
XDG_MUSIC_DIR="$HOME/Music"
XDG_PICTURES_DIR="$HOME/Pictures"
XDG_VIDEOS_DIR="$HOME/Videos"
```
I decided to do a slow test of things with `/home/chris/.config/firejail/firefox_anime.local` whitelisted entry for /Downloads REM'd out.

```
ignore apparmor
ignore noexec ${HOME}
whitelist ${HOME}/firefox
whitelist /data/chris/firefox_profiles/anime
#whitelist ${HOME}/Downloads
```
Then I did the following tests -

1. In a terminal I ran the Exec line from my .desktop file. I could not access my Downloads directory in firefox. I note that profile `firefox.profile` is not listed in the output.
```
chris@asus-roc:~>$firejail  --profile=/home/chris/.config/firejail/firefox_anime.local  /home/chris/firefox/firefox --profile /data/chris/firefox_profiles/anime --no-remote 
Reading profile /home/chris/.config/firejail/firefox_anime.local
Parent pid 27200, child pid 27201
Child process initialized in 6.86 ms

Parent is shutting down, bye...
chris@asus-roc:~>$
```

2. I reran the terminal by adding the `firefox.profile`to the command. Now my firefox sees Downloads and the other profiles are listed -
```
chris@asus-roc:~>$firejail  --profile=/home/chris/.config/firejail/firefox_anime.local --profile=firefox /home/chris/firefox/firefox --profile /data/chris/firefox_profiles/anime --no-remote 
Reading profile /home/chris/.config/firejail/firefox_anime.local
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 27528, child pid 27532
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: NVIDIA card detected, nogroups command ignored
Warning: NVIDIA card detected, nogroups command ignored
Warning: cleaning all supplementary groups
Warning: Replacing profile instead of stacking it. It is a legacy behavior that can result in relaxation of the protection. It is here as a temporary measure to unbreak the software that has been broken by switching to the stacking behavior.
Warning: Cannot confine the application using AppArmor.
Maybe firejail-default AppArmor profile is not loaded into the kernel.
As root, run "aa-enforce firejail-default" to load it.
Child process initialized in 187.83 ms

Parent is shutting down, bye...
chris@asus-roc:~>$
```
3. Based on 2 above I added the `firefox.profile` to the desktop file. But now firefox does not see Downloads directory.

```
[Desktop Entry]
Categories=Network;WebBrowser;GTK;
Encoding=UTF-8
Name=Firefox(Anime)
GenericName=Web Browser
Comment=Web Browser
TryExec=/home/chris/firefox/firefox
Exec=firejail --profile=/home/chris/.config/firejail/firefox_anime.local --profile=firefox /home/chris/firefox/firefox --profile /data/chris/firefox_profiles/anime --no-remote  %u
Icon=/home/chris/firefox/browser/chrome/icons/default/default128.png
Terminal=false
StartupNotify=true
MimeType=text/html;text/xml;application/xhtml+xml;application/vnd.mozilla.xul+xml;text/mml;application/x-xpinstall;x-scheme-handler/http;x-scheme-handler/https;x-scheme-handler/ftp;
Type=Application

Actions=PrivateBrowsing;

[Desktop Action PrivateBrowsing]
Name=New Private Browsing Window
Exec=/home/chris/firefox/firefox --private-window  %u

```
I'm getting very confused now as to why there's a different result between the terminal ran command and the .desktop file.

Help me Obi-wan.

**Reply by kzsa** [reply]
I am the frog in the swamp and not Obi... xD
I am not an expert on firejail, I just use it. It works under Suse as I suggested. I tried it yesterday.
I'm sorry, but I think you're going a bit overboard.
If you want to go deeper into the debugging, this is what recommended to read (In addition to the firejail documentation):
https://debugging.works/blog/debugging-firejail/

**Comment by therealkitman** [comment]
> B)
> Running multiple instances in parallel with different Firefox profiles (not different firejail profiles, different Firefox profiles!)

Hi @kzsa . I separate  my firefox profiles for the following reason(s).

Creating profiles via firefox's ProfileManager is fine but all those profiles referenced in `/home/user/.mozilla/firefox/profiles.ini` share the same cache in `/home/user/.cache/mozilla/firefox/`. So I make separate firefox profiles  on a second SSD. 
 
When I start firefox with `firefox --profile /path/to/my/ffprofile` ALL session data including the cache are stored in that profile.  So that makes them very portable wihout relying on `/home/user/.mozilla/firefox/profiles.ini`.

Just trying to isolate things as much as possible.

**Comment by ghost** [comment, marked as answer]
> In a terminal I ran the Exec line from my .desktop file. I could not access my Downloads directory in firefox. I note that profile firefox.profile is not listed in the output.

Well that's to be expected when you don't explicitly `whitelist ${DOWNLOADS}` (note the difference with `whitelist ${HOME}/Downloads` here!) in firefox_anime.local and don't `include firefox.profile` either.

> Based on 2 above I added the firefox.profile to the desktop file. But now firefox does not see Downloads directory.

Suggestions for de-confusement:

- rename firefox_anime.local to `firefox_anime.profile` (foo.local only makes sense when foo.profile exists and takes care of including foo.local)
- add `include /etc/firejail/firefox.profile` to firefox_anime.profile (this way you'll have all the defaults, including `whitelist ${DOWNLOADS}`)
- simplify the relevant desktop file's Exec line
    -  use firejail's `--` parameter to indicate end of firejail options; what follows are considered parameters to be past to the referenced application, in this case `/home/chris/firefox/firefox`
    - shorten the paths to firejail profiles
    - use sh -c 'foo' to ensure the correct execution of your custom command

Summing up:

```sh
$ cat ~/.config/firejail/firefox_anime.profile
ignore apparmor
ignore noexec ${HOME}
whitelist ${HOME}/firefox
whitelist /data/chris/firefox_profiles/anime
# redirect to default firefox profile
include /etc/firejail/firefox.profile
```

```sh
$ cat ~/.local/share/applications/firefox_anime.desktop
[Desktop Entry]
Categories=Network;WebBrowser;GTK;
Encoding=UTF-8
Name=Firefox (Anime)
GenericName=Web Browser
Comment=Web Browser
TryExec=/home/chris/firefox/firefox
Exec=sh -c 'firejail --profile=firefox_anime.profile -- /home/chris/firefox/firefox -P /data/chris/firefox_profiles/anime --no-remote  %u'
Icon=/home/chris/firefox/browser/chrome/icons/default/default128.png
Terminal=false
StartupNotify=true
MimeType=text/html;text/xml;application/xhtml+xml;application/vnd.mozilla.xul+xml;text/mml;application/x-xpinstall;x-scheme-handler/http;x-scheme-handler/https;x-scheme-handler/ftp;
Type=Application

Actions=PrivateBrowsing;

[Desktop Action PrivateBrowsing]
Name=New Private Browsing Window
Exec=/home/chris/firefox/firefox --private-window  %u
```

HTH

**Comment by therealkitman** [comment]
> Suggestions for de-confusement:

Brilliant. Your suggestions worked!!!!! Only change I had to make was that the `firefox_anime.profile` needed  path.

So from 
`Exec=sh -c 'firejail --profile=firefox_anime.profile -- /home/chris/firefox/firefox --profile /data/chris/firefox_profiles/anime --no-remote'`
to
`Exec=sh -c 'firejail --profile=/home/chris/.config/firejail/firefox_anime.profile -- /home/chris/firefox/firefox --profile /data/chris/firefox_profiles/anime --no-remote'`

Many thanks!! :+1:
